pkg/ruler/ruler_test.go

```diff
@@ -844,9 +899,89 @@ func TestGetRules(t *testing.T) {
 				time.Sleep(100 * time.Millisecond)
 			}
 
+			scrape := func(rID string, scrapeCh chan prometheus.Metric, doneCh chan struct{}) {
+				go func() {
+					tick := time.NewTicker(10 * time.Millisecond)
+					for {
+						select {
+						case <-tick.C:
+							if len(scrapeCh) >= 100 {
+								continue
+							}
+							m := rulerManagerMap[rID]
+							data := m.userManagerMetrics.regs.BuildMetricFamiliesPerUser()
+							data.SendSumOfCountersPerUserWithLabels(scrapeCh, m.userManagerMetrics.IterationsScheduled, "prometheus_rule_group_iterations_total", "rule_group")
+						case <-doneCh:
+							return
+						}
+					}
+				}()
+			}
+
+			waitUntil := func(rID string, user string, ruleGroupList rulespb.RuleGroupList) {
+				scrapeCh := make(chan prometheus.Metric, 100)
+				doneCh := make(chan struct{})
+				scrape(rID, scrapeCh, doneCh)
+				m := map[string]float64{}
+				for _, rgDesc := range ruleGroupList {
+					ruler := rulerAddrMap[rID]
+					if tc.sharding {
+						subRing := ruler.ring.ShuffleShard(user, tc.shuffleShardSize)
+						owns, err := instanceOwnsRuleGroup(subRing, rgDesc, validation.DisabledRuleGroups{}, ruler.lifecycler.GetInstanceAddr(), false)
+						if err == nil && owns {
+							m[rgDesc.GetName()] = 0
+						}
+					} else {
+						m[rgDesc.GetName()] = 0
+					}
+				}
+				if len(m) == 0 {
+					close(doneCh)
+					return
+				}
+				go func() {
+					defer close(doneCh)
+					for metric := range scrapeCh {
+						collected := &dto.Metric{}
+						_ = metric.Write(collected)
+						userLabel := ""
+						ruleGroupName := ""
+						for _, lp := range collected.GetLabel() {
+							if lp.GetName() == "user" && lp.GetValue() == user {
+								userLabel = lp.GetValue()
+							}
+							if lp.GetName() == "rule_group" {
+								ruleGroupName = lp.GetValue()
+							}
+						}
+						if userLabel != "" && ruleGroupName != "" {
+							for k := range m {
+								if strings.Contains(ruleGroupName, k) {
+									m[k] = m[k] + collected.Counter.GetValue()
+								}
+							}
+						}
+						done := true
+						for k := range m {
+							if m[k] < 3 {
+								done = false
+								break
+							}
+						}
+						if done {
+							return
+						}
+					}
```

Comment:
you can use labels to simplify the code a bit.
```suggestion
					outerLoop:
					for metric := range scrapeCh {
						collected := &dto.Metric{}
						_ = metric.Write(collected)
						userLabel := ""
						ruleGroupName := ""
						for _, lp := range collected.GetLabel() {
							if lp.GetName() == "user" && lp.GetValue() == user {
								userLabel = lp.GetValue()
							}
							if lp.GetName() == "rule_group" {
								ruleGroupName = lp.GetValue()
							}
						}
						if userLabel != "" && ruleGroupName != "" {
							for k := range m {
								if strings.Contains(ruleGroupName, k) {
									m[k] = m[k] + collected.Counter.GetValue()
								}
							}
						}
						for k := range m {
							if m[k] < 3 {
								continue outerLoop
							}
						}
					}
```